File: apps/pulse_app/test/timeline_query_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:pulse_protocol/pulse_wire.dart';

import 'package:pulse/features/timeline/timeline_query.dart';

TimelineEvent _ev({
  int severity = Severity.info,
  String channel = 'System',
  String category = 'Service',
  String provider = 'Service Control Manager',
  int winEventId = 7036,
  String computer = 'DESKTOP',
  String message = 'The service entered the running state.',
  String processName = '',
  bool hasProcessId = false,
  int processId = 0,
  int timestampUnixMs = 1_700_000_000_000,
  String rawXml = '',
}) {
  return TimelineEvent(
    eventId: '$channel|$winEventId',
    severity: severity,
    channel: channel,
    category: category,
    providerName: provider,
    winEventId: winEventId,
    computerName: computer,
    message: message,
    title: 'Title',
    summary: 'Summary',
    processName: processName,
    hasProcessId: hasProcessId,
    processId: processId,
    timestampUnixMs: timestampUnixMs,
    rawXml: rawXml,
  );
}

void main() {
  group('TimelineQuery', () {
    test('empty query matches everything', () {
      expect(const TimelineQuery().matches(_ev()), isTrue);
      expect(const TimelineQuery().isActive, isFalse);
    });

    test('severity errors filter', () {
      const q = TimelineQuery(severity: TimelineSeverityFilter.errors);
      expect(q.matches(_ev(severity: Severity.error)), isTrue);
      expect(q.matches(_ev(severity: Severity.critical)), isTrue);
      expect(q.matches(_ev(severity: Severity.warning)), isFalse);
      expect(q.isActive, isTrue);
    });

    test('source application filter', () {
      const q = TimelineQuery(source: TimelineSourceFilter.application);
      expect(q.matches(_ev(channel: 'Application')), isTrue);
      expect(q.matches(_ev(channel: 'System')), isFalse);
    });

    test('category crash filter', () {
      const q = TimelineQuery(category: TimelineCategoryFilter.crash);
      expect(q.matches(_ev(category: 'Crash')), isTrue);
      expect(q.matches(_ev(category: 'Service')), isFalse);
    });

    test('provider filter is case-insensitive', () {
      const q = TimelineQuery(providerContains: 'kernel-power');
      expect(
        q.matches(_ev(provider: 'Microsoft-Windows-Kernel-Power')),
        isTrue,
      );
      expect(q.matches(_ev(provider: 'Service Control Manager')), isFalse);
    });

    test('event id exact match', () {
      const q = TimelineQuery(eventIdEquals: '41');
      expect(q.matches(_ev(winEventId: 41)), isTrue);
      expect(q.matches(_ev(winEventId: 42)), isFalse);
      expect(const TimelineQuery(eventIdEquals: 'abc').matches(_ev()), isFalse);
    });

    test('process name and pid', () {
      final withProc = _ev(
        processName: 'explorer.exe',
        hasProcessId: true,
        processId: 4242,
      );
      expect(
        const TimelineQuery(processContains: 'EXPLORER').matches(withProc),
        isTrue,
      );
      expect(
        const TimelineQuery(processContains: '4242').matches(withProc),
        isTrue,
      );
      expect(
        const TimelineQuery(processContains: 'chrome').matches(withProc),
        isFalse,
      );
    });

    test('computer filter', () {
      const q = TimelineQuery(computerContains: 'desk');
      expect(q.matches(_ev(computer: 'DESKTOP-1')), isTrue);
      expect(q.matches(_ev(computer: 'LAPTOP')), isFalse);
    });

    test('date range last 15 minutes', () {
      const now = 1_700_000_360_000;
      final q = TimelineQuery(
        dateRange: TimelineDateRangeFilter.last15Minutes,
        nowUnixMs: now,
      );
      expect(
        q.matches(_ev(timestampUnixMs: now - 10 * 60 * 1000)),
        isTrue,
      );
      expect(
        q.matches(_ev(timestampUnixMs: now - 20 * 60 * 1000)),
        isFalse,
      );
      expect(q.matches(_ev(timestampUnixMs: 0)), isFalse);
    });

    test('date range last hour', () {
      const now = 1_700_000_360_000;
      final q = TimelineQuery(
        dateRange: TimelineDateRangeFilter.lastHour,
        nowUnixMs: now,
      );
      expect(
        q.matches(_ev(timestampUnixMs: now - 30 * 60 * 1000)),
        isTrue,
      );
      expect(
        q.matches(_ev(timestampUnixMs: now - 2 * 60 * 60 * 1000)),
        isFalse,
      );
      expect(q.matches(_ev(timestampUnixMs: 0)), isFalse);
    });

    test('search covers provider eventId computer message xml process pid', () {
      final e = _ev(
        provider: 'Application Error',
        winEventId: 1000,
        computer: 'WORKPC',
        message: 'Faulting application name: notepad.exe',
        processName: 'notepad.exe',
        hasProcessId: true,
        processId: 99,
        rawXml: '<Event><Data>UniqueXmlToken</Data></Event>',
      );
      expect(const TimelineQuery(searchQuery: 'application error').matches(e), isTrue);
      expect(const TimelineQuery(searchQuery: '1000').matches(e), isTrue);
      expect(const TimelineQuery(searchQuery: 'workpc').matches(e), isTrue);
      expect(const TimelineQuery(searchQuery: 'faulting').matches(e), isTrue);
      expect(const TimelineQuery(searchQuery: 'uniquexmltoken').matches(e), isTrue);
      expect(const TimelineQuery(searchQuery: 'notepad').matches(e), isTrue);
      expect(const TimelineQuery(searchQuery: '99').matches(e), isTrue);
      expect(const TimelineQuery(searchQuery: 'no-such').matches(e), isFalse);
    });

    test('AND combines dedicated filters with search', () {
      final e = _ev(provider: 'Kernel-Power', winEventId: 41, category: 'Power');
      final q = const TimelineQuery(
        category: TimelineCategoryFilter.power,
        eventIdEquals: '41',
        searchQuery: 'kernel',
      );
      expect(q.matches(e), isTrue);
      expect(
        q.copyWith(eventIdEquals: '42').matches(e),
        isFalse,
      );
    });
  });
}

File: apps/pulse_mcp/test/timeline.query.test.ts
import { describe, expect, it } from "vitest";

import { emptyTimelineEvent } from "../src/timeline/types.js";
import { filterTimelineEvents, matchesTimelineEvent } from "../src/timeline/query.js";
import { mapTimelineEvent } from "../src/timeline/mappers.js";

describe("timeline query (Flutter parity)", () => {
  it("filters by severity error/critical", () => {
    const err = emptyTimelineEvent();
    err.severity = 3;
    err.title = "Crash";
    const info = emptyTimelineEvent();
    info.severity = 1;
    info.title = "Info";
    const matched = filterTimelineEvents([err, info], {
      severity: ["error", "critical"],
    });
    expect(matched).toHaveLength(1);
    expect(matched[0]!.title).toBe("Crash");
  });

  it("filters channel security vs system", () => {
    const sec = emptyTimelineEvent();
    sec.channel = "Security";
    const sys = emptyTimelineEvent();
    sys.channel = "System";
    expect(matchesTimelineEvent(sec, { channel: "security" })).toBe(true);
    expect(matchesTimelineEvent(sys, { channel: "security" })).toBe(false);
  });

  it("keyword search covers provider and message", () => {
    const e = emptyTimelineEvent();
    e.providerName = "Microsoft-Windows-Kernel-Power";
    e.message = "The system is entering sleep.";
    expect(matchesTimelineEvent(e, { keyword: "kernel-power" })).toBe(true);
    expect(matchesTimelineEvent(e, { keyword: "sleep" })).toBe(true);
    expect(matchesTimelineEvent(e, { keyword: "nomatch" })).toBe(false);
  });

  it("maps event without rawXml by default", () => {
    const e = emptyTimelineEvent();
    e.eventId = "abc";
    e.timestampIso = "2026-08-02T00:00:00.000Z";
    e.severity = 2;
    e.title = "Warn";
    e.summary = "Something";
    e.rawXml = "<Event/>";
    const mapped = mapTimelineEvent(e);
    expect(mapped.rawXml).toBeUndefined();
    expect(mapped.severity).toBe("warning");
    const withRaw = mapTimelineEvent(e, { includeRaw: true });
    expect(withRaw.rawXml).toBe("<Event/>");
  });
});
